src/hotspot/share/gc/shared/memAllocator.hpp

```diff
@@ -118,4 +118,38 @@ class ClassAllocator: public MemAllocator {
   virtual oop initialize(HeapWord* mem) const;
 };
 
+// Manages a scope where a failed heap allocation results in
+// suppression of JVMTI "resource exhausted" events and
+// throwing a shared, backtrace-less OOME instance.
+// Used for OOMEs that will not be propagated to user code.
+class SandboxedOOMEMark: public StackObj {
+ private:
+  bool _outer;
+  JavaThread* _thread;
+
+ public:
+  SandboxedOOMEMark(JavaThread* thread) {
+    if (thread != nullptr) {
+      _outer = thread->in_sandboxed_oome_mark();
+      thread->set_in_sandboxed_oome_mark(true);
+      _thread = thread;
+    } else {
+      _outer = false;
+      _thread = nullptr;
+    }
```

Comment:
It isn't obvious to me how this part is intended to be used. I see it ties back to the retryable allocation "activate" mode, but I'm unclear what that means as well.

Reply:
"this part" means the "else branch" which means the null receiving constructor. Yeah that whole "null_on_fail" thing had me a bit perplexed and I see there is now a JBS issue filed. to kill it off as we always want null-on-fail.